Add subscription deferral for Google play

antonio0 opened the pull request:
Implemented this functionality: https://developers.google.com/android-publisher/api-ref/purchases/subscriptions/defer

I know for sure that Apple doesn't provide an analogous functionality, not sure about the other providers though.

antonio0 committed:
Added defer subscription for google play

antonio0 committed:
Added defer subscription for google play

antonio0 committed:
fix

antonio0 committed:
fix

antonio0 committed:
passing deferralInfo

antonio0 committed:
passing deferralInfo

antonio0 committed:
fix lint

antonio0 committed:
fix lint

ronkorving commented on README.md:
````diff
@@ -57,6 +57,21 @@ iap.cancelSubscription("google", payment, function (error, response) {
+```javascript
+var deferralInfo = {
+	expectedExpiryTimeMillis: 1546616722237,
+  desiredExpiryTimeMillis: 1547716722237,
````
Indentation seems to have gone wrong here.

ronkorving commented:
I would love some feedback from other Google Play users (as I am not one).
Can anyone weigh in on the approach taken here?

antonio0 committed:
fixed indentation

antonio0 committed:
fixed indentation

antonio0 commented:
At our company we're running this code in production since January and have processed a have used this function a few thousand times (to delay some subscription payments). It's been stable and reliable so far, never had a problem with it. 

It would be great to see it merged 🚢

ronkorving commented:
@antonio0 Could you update this to use Google API v3?

antonio0 committed:
Merge branch 'master' into google-defer

VK432 committed:
google deferral (#1)

VK432 committed:
google deferral (#1)

antonio0 commented:
> @antonio0 Could you update this to use Google API v3?

@ronkorving just updated now to v3

antonio0 committed:
Bug fix

antonio0 committed:
Bug fix

antonio0 committed:
lint fix

antonio0 committed:
lint fix

ronkorving commented on index.js:
```diff
@@ -67,3 +67,39 @@ exports.cancelSubscription = function (platform, payment, cb) {
+
+	if (!engine.deferSubscription) {
+		return syncError(new Error(`Platform ${platform
+		} does not have deferSubscription method`));
```
Was the line too long for lint's liking?

superandrew213 committed:
[Apple] Check for cancellation_date_ms first (#50)

[Apple] Check for cancellation_date_ms first

If cancellation date is available it means that Apple Customer Support refunded subscription payment.

From Apple:
To check whether a purchase has been canceled by Apple Customer Support, look for the
Cancellation Date field in the receipt. If the field contains a date, regardless of the
subscription’s expiration date, the purchase has been canceled. With respect to providing
content or service, treat a canceled transaction the same as if no purchase had ever been made.

justinpage committed:
Update available response status codes from apple

Referencing the following documentation:

https://developer.apple.com/documentation/appstorereceipts/status

We can see that 21009 and 21010 are available codes we should handle.

ronkorving committed:
Bump ESLint

ronkorving committed:
Release 1.1.1

justinpage commented:
@antonio0 With nine commits might make sense to squash?

faboyds committed:
Created verifyPayment alternative function to return Promise (#69)

verifyPayment and cancelSubscription now return promises if no callback is passed

antonio0 committed:
Merge branch 'google-defer' of https://github.com/emma-app/node-iap into google-defer
